File: pyround1/1022.从根到叶的二进制数之和.py
#
# @lc app=leetcode.cn id=1022 lang=python3
#
# [1022] 从根到叶的二进制数之和
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sumRootToLeaf(self, root: TreeNode) -> int:
        if not root :
            return 0
        res = []
        q = []
        q.append([root, '0b'])
        while q:
            node, tmp = q.pop(0)
            tmp += str(node.val)

            if node.left:
                q.append([node.left, tmp])
            if node.right:
                q.append([node.right, tmp])
            
            if not node.left and not node.right:
                res.append(tmp)
        
        return sum(int(bin_str, 2) for bin_str in res)
        
# @lc code=end

File: 1022.从根到叶的二进制数之和.cpp
/*
 * @lc app=leetcode.cn id=1022 lang=cpp
 *
 * [1022] 从根到叶的二进制数之和
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int sum = 0;
    int sumRootToLeaf(TreeNode* root) {
        helper(root, 0);
        return sum;
    }
    void helper(TreeNode* root, int val){
        if(!root)
            return ;
        if(!root->left and !root->right){
            sum += (val << 1 | root->val);
            return ;
        }
        helper(root->left, val<< 1 | root->val);
        helper(root->right, val << 1 | root->val);
        
    }
};
// @lc code=end

